cluster/swarm/cluster.go

```diff
@@ -177,6 +182,74 @@ func (c *Cluster) CreateContainer(config *cluster.ContainerConfig, name string,
 	return container, err
 }
 
+// setOSTypeConstraint chooses an engine and leverages its /distribution
+// endpoint to determine a list of compatible Platforms for the image. it then
+// adds an ostype constraint for the valid OS types. If an ostype constraint
+// already exists on container, then this will not overwrite it.
+func (c *Cluster) setOSTypeConstraint(config *cluster.ContainerConfig, authConfig *types.AuthConfig) error {
+	// first, check if there is an existing ostype constraint. if so, leave this
+	// alone and take no action.
+	constraints := config.Constraints()
+	for _, constraint := range constraints {
+		if strings.Contains(constraint, "ostype") {
+			return nil
+		}
+	}
+
+	// now that we know we have to set an os constraint, choose an engine at
+	// random. any engine should theoretically be able to contact the registry
+	engine, err := c.RANDOMENGINE()
+	if err != nil {
+		return err
+	}
+
+	// now call the corresponding method on this engine
+	platforms, err := engine.GetImagePlatforms(config, authConfig)
+	if err != nil {
+		return err
+	}
+
+	// now extract the OSes. use a map to deduplicate, in case the image is
+	// available on several architectures with the same platform.
+	ostypes := map[string]struct{}{}
+	for _, p := range platforms {
+		ostypes[p.OS] = struct{}{}
```

Comment:
If you are depending on the [OS](https://github.com/moby/moby/blob/8e610b2b55bfd1bfa9436ab110d311f5e8a74dcb//vendor/github.com/opencontainers/image-spec/specs-go/v1/descriptor.go#L51) field alone, it may not work for Windows as Windows images are sensitive to OS versions. If you try to run a 2016 image on a 2019 OS or vice versa, Windows HCS will reject the image and the container won't run. To make this work correctly on Windows, the OS versions need to be matched as well (besides OS type)

Reply:
I don't see where the engine reports OS version, though. Swarm relies on the output of `docker info` to discover information like the OS, and `info` does have a particular field that appears to map 1:1 to the OSVersion field in the Platform.

Reply:
Then I think this fix is sufficient if we can document the expected behaviors for the customers: 1. we recommend the customers to explicitly specify OS-info in the image manifest (and swarm matches it correctly). 2. we make error messages as useful as possible when mismatch happens.